spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarTemplate.java

```diff
@@ -192,13 +204,19 @@ private CompletableFuture<MessageId> doSendAsync(@Nullable String topic, T messa
 			observation.start();
 			Producer<T> producer = prepareProducerForSend(topicName, message, schema, encryptionKeys,
 					producerCustomizer);
-			TypedMessageBuilder<T> messageBuilder = producer.newMessage().value(message);
-			if (typedMessageBuilderCustomizer != null) {
-				typedMessageBuilderCustomizer.customize(messageBuilder);
+			TypedMessageBuilder<T> messageBuilder;
+			try {
+				messageBuilder = producer.newMessage().value(message);
+				if (typedMessageBuilderCustomizer != null) {
+					typedMessageBuilderCustomizer.customize(messageBuilder);
+				}
+				// propagate props to message
+				senderContext.properties().forEach(messageBuilder::property);
+			}
+			catch (Exception e) {
+				ProducerUtils.closeProducerAsync(producer, this.logger);
```

Comment:
Oh, what brought this up?

Reply:
If the message type doesn't match the Producer one (eg. the message type is Foo.class and the producer is Schema.BYTES), `producer.newMessage().value(message)` throws a cast exception and the producer is not closed.